### DISCUSSION TITLE
Touch controls using trame-rca

### DISCUSSION THREAD

**Discussion by mashirafarid** [Kitware/trame#615] [Trame/ParaView]
Hi, so progress on my paraview/trame web app has been going pretty well, but I noticed that if I run my app on say an iPad, touch controls like pinch to zoom in/out don't work when using VtkRemoteView. It was mentioned in #362 that this functionality will be added to `trame-rca`, so just wanted to check if this has been implemented yet, especially for Paraview? I did try following this [example](https://github.com/Kitware/trame-rca/blob/master/examples/00_cone/app.py) to add `trame-rca` to my app, but I can't seem to get it working.

**Comment by jourdain** [comment]
`trame-rca` could indeed work, but it will be up to the implementation/binding to properly convert those touch events to the VTK event type. A quick test on ParaView Async which use trame-rca show that it is not working out of the box.

In other words, there might still be some work to be done to get that working. But did you try the vtk example from rca? Does that one work with touch event or not?

If you need help pushing that forward, you may want to get some [professional support](https://www.kitware.com/trame/).

**Reply by mashirafarid** [reply]
My current set up is that I run the app on my MacBook using the `--host` argument, then use my iPad to interact with the app. I tried running the vtk rca example like this but for some reason I can't seem to access it via the iPad, only via the MacBook. I did screen mirror my MacBook onto the iPad and from this found that pinch to zoom works, but rotating the cone with one finger doesn't work; however this seems to be an issue with the screen mirroring itself and not `trame-rca`. To confirm this, I will need to find a way to access the example using the iPad, is there a way to run it so devices on my local network can access it?

**Reply by jourdain** [reply]
Did you use `--host 0.0.0.0`? That should allow your ipad to connect to your computer over the network.

**Reply by mashirafarid** [reply]
So initially I was using `--host [macbook IP address]` which wasn't working, but I tried again today and it started working, not sure why. Anyways, I can rotate the cone, but pinching gives the following error:
```shell
2024-10-19 20:09:28.342 (  63.079s) [          62BEBB]vtkRemoteInteractionAda:279    ERR| Skipping Event 
[json.exception.out_of_range.403] key 'positions' not found
``` 
Is this what you meant when you said that it will be up to the implementation/binding to properly convert those touch events to the VTK event type?

Although I don't think I'll need three-finger gestures, I tried it and get messages like this:
```shell
JS Error => error: TypeError: undefined is not an object (evaluating 'm[1].x')
```

**Reply by jourdain** [reply]
I guess more work needs to be done to enable full support for touch, but at least the infrastructure is there.
